osu.Game/Beatmaps/Drawables/DifficultyRetriever.cs

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System.Collections.Generic;
+using System.Threading;
+using osu.Framework.Allocation;
+using osu.Framework.Bindables;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Game.Rulesets;
+using osu.Game.Rulesets.Mods;
+
+namespace osu.Game.Beatmaps.Drawables
+{
+    /// <summary>
+    /// A component solely responsible for calculating difficulty in the background.
+    /// Intended for use with <see cref="DelayedLoadWrapper"/> to only run processing when usage is on-screen.
+    /// </summary>
+    public class DifficultyRetriever : Component
+    {
+        /// <summary>
+        /// The bindable star difficulty.
+        /// </summary>
+        public IBindable<StarDifficulty> StarDifficulty => starDifficulty;
+
+        private readonly Bindable<StarDifficulty> starDifficulty = new Bindable<StarDifficulty>();
+
+        private readonly IBeatmapInfo beatmapInfo;
+
+        private readonly IRulesetInfo? ruleset;
+        private readonly IReadOnlyList<Mod>? mods;
+
+        private readonly CancellationTokenSource difficultyCancellation = new CancellationTokenSource();
+
+        [Resolved]
+        private BeatmapDifficultyCache difficultyCache { get; set; } = null!;
+
+        /// <summary>
+        /// Construct a difficulty retriever that tracks the current ruleset / mod selection.
+        /// </summary>
+        /// <param name="beatmapInfo">The beatmap to use for calculation.</param>
+        public DifficultyRetriever(IBeatmapInfo beatmapInfo)
+        {
+            this.beatmapInfo = beatmapInfo;
+        }
+
+        /// <summary>
+        /// Construct a difficulty retriever that is calculated only once for the specified ruleset / mod combination.
+        /// This will not track global ruleset and mod changes.
+        /// </summary>
+        /// <param name="beatmapInfo">The beatmap to use for calculation.</param>
+        /// <param name="ruleset">The ruleset to use for calculation.</param>
+        /// <param name="mods">The mods to use for calculation.</param>
+        public DifficultyRetriever(IBeatmapInfo beatmapInfo, IRulesetInfo ruleset, IReadOnlyList<Mod> mods)
+        {
+            this.beatmapInfo = beatmapInfo;
+            this.ruleset = ruleset;
+            this.mods = mods;
+        }
```

Comment:
I'm not sure if this is worth keeping around. It is not used anywhere right now. It is also the only usage of:

https://github.com/ppy/osu/blob/f8830c6850128456266c82de83273204f8b74ac0/osu.Game/Beatmaps/BeatmapDifficultyCache.cs#L104-L117

Reply:
I think it'll be useful for cases like queued up `DrawableRoomPlaylistItem` which are not currently selected but need to display the correct SR icon in the item.

Reply:
Is now gone anyway.